src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IListOfTConverter.cs

```diff
@@ -12,9 +13,18 @@ internal sealed class IListOfTConverter<TCollection, TElement>
         : IEnumerableDefaultConverter<TCollection, TElement>
         where TCollection : IList<TElement>
     {
+        // Used to store elements when TCollection is ValueType
+        [MaybeNull]
+        private TCollection _listValues;
+
         protected override void Add(in TElement value, ref ReadStack state)
         {
-            ((TCollection)state.Current.ReturnValue!).Add(value);
+            if (_listValues != null)
```

Comment:
Something like this should work instead:
```cs
            TCollection collection = (TCollection)state.Current.ReturnValue!;
            collection.Add(value);
            if (IsValueType)
            {
                state.Current.ReturnValue = collection;
            };
```

Reply:
Thank @steveharter  I have just updated my PR.